Adding details document about confirmation pages frontend refactoring (#17742)
MetaMask/metamask-extension: docs/refactoring/confirmation-pages-routing/README.md

# Refactoring - Confirmation pages routing

This document details how routing to confirmation pages is currently done and the proposed improvements in routing.

## Current flow

The current flow of routing to confirmation pages is un-necessarily complicated and have issues.

![Confirmation Pages Routing - Current](https://raw.githubusercontent.com/MetaMask/metamask-extension/develop/docs/refactoring/confirmation-pages-routing/current.png)

- There are 2 ways in which confirmation pages can be opened:
  1. User triggers send flow from within Metamask
     - If the user triggers the send flow from within MetaMask and selects the recipient and amount on the send screen, an unapproved transaction is created in the background and the user is redirected to the **/confirm-transaction** route.
  2. DAPP sends request to Metamask

>      - If DAPP sends request to Metamask an un-approved transaction or signature request is created in background and UI is triggered open (if it is not already open).

Reviewer: ```suggestion
     - If DAPP sends request to Metamask an unapproved transaction or signature request is created in background and UI is triggered open (if it is not already open).
```

Author: I fixed it in this following PR: https://github.com/MetaMask/metamask-extension/pull/17821

# Refactoring - Confirmation pages routing

This document details how routing to confirmation pages is currently done and the proposed improvements in routing.

## Current flow

The current flow of routing to confirmation pages is un-necessarily complicated and have issues.

![Confirmation Pages Routing - Current](https://raw.githubusercontent.com/MetaMask/metamask-extension/develop/docs/refactoring/confirmation-pages-routing/current.png)

- There are 2 ways in which confirmation pages can be opened:
  1. User triggers send flow from within Metamask

>      - If the user trigger send flow from within Metamask, user selects receipient and amount and on send screen, at this point an un-approved transaction is created in the background and user is re-directed to **/confirm-transaction** route.

Reviewer 2: ```suggestion
     - If the user triggers the send flow from within MetaMask and selects the recipient and amount on the send screen, an unapproved transaction is created in the background and the user is redirected to the **/confirm-transaction** route.
```

| /confirm-transaction/${id}/deploy-contract      | pages/confirm-deploy-contract        |
| /confirm-transaction/${id}/send-ether           | pages/confirm-send-ether             |
| /confirm-transaction/${id}/send-token           | pages/confirm-send-token             |
| /confirm-transaction/${id}/approve              | pages/confirm-approve                |
| /confirm-transaction/${id}/set-approval-for-all | pages/confirm-approve                |
| /confirm-transaction/${id}/transfer-from        | pages/confirm-token-transaction-base |
| /confirm-transaction/${id}/safe-transfer-from   | pages/confirm-token-transaction-base |
| /confirm-transaction/${id}/token-method         | pages/confirm-contract-interaction   |
| /confirm-transaction/${id}/signature-request    | pages/confirm-signature-request.js   |

## Proposed flow

The proposed routing of confirmation pages looks like.

![Confirmation Pages Routing - Proposed](https://raw.githubusercontent.com/MetaMask/metamask-extension/develop/docs/refactoring/confirmation-pages-routing/proposed.png)

- There are 2 ways in which confirmation pages can be opened:
  1. User triggers send flow from within Metamask

>      - If the user trigger send flow from within Metamask, user selects receipient and amount and on send screen, at this point an un-approved transaction is created in background and user is re-directed to specific transaction route **/confirm-transaction/${id}/XXXX** depending on transaction.

Reviewer 2: ```suggestion
     - [changed] If the user triggers the send flow from within MetaMask and selects the recipient and amount on the send screen, an unapproved transaction is created in the background and the user is redirected to a specific transaction route, **/confirm-transaction/${id}/XXXX**, depending on the transaction type.
```

  2. DAPP sends request to Metamask
     - If DAPP send request to Metamask an un-approved transaction or signature request is created in background and UI is triggered to open (if it is not already open).

>      - The router find un-approved transaction in state and re-route to **/confirm-transaction**.

Reviewer 2: It was a little tricky for me to understand the changes here. I think I understand now after looking at the charts. 

Maybe we can elaborate a bit more here. 

proposed: 
```suggestion
     - [changed] Instead of rendering `pages/home`,  `pages/routes` finds the unapproved transaction in state and reroutes to **/confirm-transaction**.
```
